File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        int prev = nums[0];
        int prev2 = 0;

        for(int i = 0; i < n; i++){
            int take = nums[i];
            if(i > 1){
                take += prev2;
            }
            int not_take = 0+prev;
            int curri = max(take, not_take);
            prev2 = prev;
            prev = curri;
        }
        return prev;
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def helper(self, idx, nums, dp):
        if idx < 0:
            return 0
        if idx == 0:
            return nums[idx]
        if dp[idx] != -1:
            return dp[idx]
        
        take = nums[idx] + self.helper(idx - 2, nums, dp)
        not_take = self.helper(idx - 1, nums, dp)
        
        dp[idx] = max(take, not_take)
        return dp[idx]

    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1] * n
        return self.helper(n-1, nums, dp)